File: sudoku.py
import random
import time

class Sudoku:
    def __init__(self):
        self.board = [[0 for _ in range(9)] for _ in range(9)]
        self.solution = [[0 for _ in range(9)] for _ in range(9)]
        
    def is_valid(self, board, row, col, num):
        # Check row
        for x in range(9):
            if board[row][x] == num:
                return False
        
        # Check column
        for x in range(9):
            if board[x][col] == num:
                return False
        
        # Check 3x3 box
        start_row = row - row % 3
        start_col = col - col % 3
        for i in range(3):
            for j in range(3):
                if board[i + start_row][j + start_col] == num:
                    return False
        
        return True
    
    def solve(self, board):
        for i in range(9):
            for j in range(9):
                if board[i][j] == 0:
                    for num in range(1, 10):
                        if self.is_valid(board, i, j, num):
                            board[i][j] = num
                            if self.solve(board):
                                return True
                            board[i][j] = 0
                    return False
        return True
    
    def generate_puzzle(self, difficulty=40):
        # Generate complete solution
        self.fill_board()
        self.solution = [row[:] for row in self.board]
        
        # Remove numbers based on difficulty
        cells_to_remove = difficulty
        while cells_to_remove > 0:
            row = random.randint(0, 8)
            col = random.randint(0, 8)
            if self.board[row][col] != 0:
                self.board[row][col] = 0
                cells_to_remove -= 1
    
    def fill_board(self):
        numbers = list(range(1, 10))
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    random.shuffle(numbers)
                    for num in numbers:
                        if self.is_valid(self.board, i, j, num):
                            self.board[i][j] = num
                            if self.solve(self.board):
                                break
                            self.board[i][j] = 0
                    else:
                        return False
        return True
    
    def print_board(self):
        for i in range(9):
            if i % 3 == 0 and i != 0:
                print("------+-------+------")
            for j in range(9):
                if j % 3 == 0 and j != 0:
                    print("| ", end="")
                print(self.board[i][j] if self.board[i][j] != 0 else ".", end=" ")
            print()
    
    def is_complete(self):
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    return False
        return True
    
    def check_solution(self):
        return self.board == self.solution

File: sudoku.js
class Sudoku {
    constructor() {
        this.board = Array(9).fill().map(() => Array(9).fill(0));
        this.solution = Array(9).fill().map(() => Array(9).fill(0));
    }

    isValid(board, row, col, num) {
        // Check row
        for (let x = 0; x < 9; x++) {
            if (board[row][x] === num) return false;
        }
        
        // Check column
        for (let x = 0; x < 9; x++) {
            if (board[x][col] === num) return false;
        }
        
        // Check 3x3 box
        const startRow = row - row % 3;
        const startCol = col - col % 3;
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                if (board[i + startRow][j + startCol] === num) return false;
            }
        }
        
        return true;
    }

    solve(board) {
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (board[i][j] === 0) {
                    for (let num = 1; num <= 9; num++) {
                        if (this.isValid(board, i, j, num)) {
                            board[i][j] = num;
                            if (this.solve(board)) return true;
                            board[i][j] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    generatePuzzle(difficulty = 40) {
        this.fillBoard();
        this.solution = this.board.map(row => [...row]);
        
        let cellsToRemove = difficulty;
        while (cellsToRemove > 0) {
            const row = Math.floor(Math.random() * 9);
            const col = Math.floor(Math.random() * 9);
            if (this.board[row][col] !== 0) {
                this.board[row][col] = 0;
                cellsToRemove--;
            }
        }
    }

    fillBoard() {
        const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9];
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (this.board[i][j] === 0) {
                    const shuffled = numbers.sort(() => Math.random() - 0.5);
                    for (const num of shuffled) {
                        if (this.isValid(this.board, i, j, num)) {
                            this.board[i][j] = num;
                            if (this.solve(this.board)) break;
                            this.board[i][j] = 0;
                        }
                    }
                }
            }
        }
    }

    printBoard() {
        for (let i = 0; i < 9; i++) {
            if (i % 3 === 0 && i !== 0) console.log("------+-------+------");
            let row = "";
            for (let j = 0; j < 9; j++) {
                if (j % 3 === 0 && j !== 0) row += "| ";
                row += (this.board[i][j] || ".") + " ";
            }
            console.log(row);
        }
    }

    isComplete() {
        return this.board.every(row => row.every(cell => cell !== 0));
    }

    checkSolution() {
        return JSON.stringify(this.board) === JSON.stringify(this.solution);
    }
}

module.exports = Sudoku;